Mathlib/Tactic/Bound/Attribute.lean

```diff
@@ -0,0 +1,138 @@
+  match e with
+  | ~q(@OfNat.ofNat.{u} _ (nat_lit 0) $i) => return true
+  | _ => return false
+
+/-- Map the arguments of an inequality expression to a score -/
+def ineqPriority (a b : Q($α)) : MetaM ℕ := do
+  return if (← isZero a) || (← isZero b) then 1 else 10
+
+/-- Map a hypothesis type to a score -/
+partial def hypPriority (hyp : Q(Prop)) : MetaM ℕ := do
+  match hyp with
+    -- Conjunctions add scores
+    | ~q($a ∧ $b) => pure <| (← hypPriority a) + (← hypPriority b)
+    -- Guessing (disjunction) gets a big penalty
+    | ~q($a ∨ $b) => pure <| 100 + (← hypPriority a) + (← hypPriority b)
+    -- Inequalities get score 1 if they contain zero, 10 otherwise
+    | ~q(@LE.le _ $i $a $b) => ineqPriority a b
+    | ~q(@LT.lt _ $i $a $b) => ineqPriority a b
+    | ~q(@GE.ge _ $i $b $a) => ineqPriority a b
+    | ~q(@GT.gt _ $i $b $a) => ineqPriority a b
+    -- Assume anything else is non-relevant
+    | _ => pure 0
+
+/-- Map a type to a score -/
+def typePriority (decl : Lean.Name) (type : Lean.Expr) : MetaM ℕ :=
+  Lean.Meta.forallTelescope type fun xs t ↦ do
+    checkResult t
+    xs.foldlM (fun (t : ℕ) x ↦ do return t + (← argPriority x)) 0
+  where
+  /-- Score the type of argument `x` -/
+  argPriority (x : Lean.Expr) : MetaM ℕ := do
+    hypPriority (← Lean.Meta.inferType x)
+  /-- Insist that our conclusion is an inequality -/
+  checkResult (t : Q(Prop)) : MetaM Unit := do match t with
+    | ~q(@LE.le _ $i $a $b) => return ()
+    | ~q(@LT.lt _ $i $a $b) => return ()
+    | ~q(@GE.ge _ $i $b $a) => return ()
+    | ~q(@GT.gt _ $i $b $a) => return ()
+    | _ => throwError (f!"`{decl}` has invalid type `{type}` as a 'bound' lemma: \
+                          it should be an inequality")
+
+/-- Map a theorem decl to a score (0 means `norm apply`, `0 <` means `safe apply`) -/
+def declPriority (decl : Lean.Name) : Lean.MetaM ℕ := do
+  match (← Lean.getEnv).find? decl with
+    | some info => do
+        typePriority decl info.type
+    | none => throwError "unknown declaration {decl}"
+
+/-- Map a score to either `norm apply` or `safe apply <priority>` -/
+def scoreToConfig (decl : Lean.Name) (score : ℕ) : Aesop.Frontend.RuleConfig :=
+  let (phase, priority) := match score with
+    | 0 => (Aesop.PhaseName.norm, 0)  -- No hypotheses: this rule closes the goal immediately
+    | s => (Aesop.PhaseName.safe, s)
+  { term? := some (Lean.mkIdent decl)
+    phase? := phase
+    priority? := some (Aesop.Frontend.Priority.int priority)
+    builder? := some (.regular .apply)
+    builderOptions := {}
+    ruleSets := ⟨#[`Bound]⟩ }
+
+/-- Register a lemma as an `apply` rule for the `bound` tactic.
+
+A lemma is appropriate for `bound` if it proves an inequality using structurally simpler
+inequalities, "recursing" on the structure of the expressions involved, assuming positivity or
+nonnegativity where useful. Examples include
+1. `gcongr`-like inequalities such as `f x ≤ f y` where `f` is monotone
```

Comment:
So to be clear: *every* lemma tagged `gcongr` should also be tagged `bound`?

Reply:
Sorry, should only be inequalities with `≤`; I believe `gcongr` is more general. I will make this very explicit.

I could try to make `bound` automatically look into `gcongr` inequalities using `≤` if we want, but I'd prefer to delay that to a later PR.